recipes/FakeItEasy.Recipes.CSharp/CapturingArguments.cs

```diff
@@ -0,0 +1,183 @@
+                capturedMessage.That.Contains("add"),
+                An<IEnumerable<int>>._))
+            .DoesNothing();
+
+        var calculator = new Calculator(logger);
+
+        // Act
+        calculator.Add([1, 2, 3, 4]);
+        calculator.Square(7);
+        calculator.Add([8, 9]);
+
+        // Assert
+        capturedMessage.Values.Should().BeEquivalentTo(
+            "about to add",
+            "about to add");
+        //// --8<-- [end:ConstrainedCapture]
+    }
+
+    [Fact]
+    public void NaivelyCaptureMutatedList()
+    {
+        var test = () =>
+        {
+            //// --8<-- [start:NaivelyCaptureMutatedList]
+            // Arrange
+            var capturedSummands = A.Captured<IEnumerable<int>>();
+
+            var logger = A.Fake<IListLogger>();
+            A.CallTo(() => logger.Log(
+                    "about to add",
+                    capturedSummands._))
+                .DoesNothing();
+
+            var calculator = new Calculator(logger);
+
+            // Act
+            List<int> summands = [1, 2, 3, 4];
+            calculator.Add(summands); // capturedSummands captures summands
+            summands.RemoveAt(0);
+            calculator.Add(summands); // captures summands again - same instance
+
+            // Assert
+            // passes:
+            capturedSummands.Values[1].Should().BeEquivalentTo(2, 3, 4);
+
+            // fails - summands contains only 2, 3, and 4:
+            capturedSummands.Values[0].Should().BeEquivalentTo(1, 2, 3, 4);
+            //// --8<-- [end:NaivelyCaptureMutatedList]
+        };
+
+        test.Should().ThrowExactly<AssertionFailedException>()
+            .WithMessage(@"*but {2, 3, 4}*contains 1 item(s) less than*{1, 2, 3, 4}*");
+    }
+
+    [Fact]
+    public void CaptureCopiedMutatedList()
+    {
+        //// --8<-- [start:CaptureCopiedMutatedList]
+        // Arrange
+        var capturedSummands = A.Captured<IEnumerable<int>>(l => l.ToList());
+
+        var logger = A.Fake<IListLogger>();
+        A.CallTo(() => logger.Log(
+                "about to add",
+                capturedSummands._))
+            .DoesNothing();
+
+        var calculator = new Calculator(logger);
+
+        // Act
+        List<int> summands = [1, 2, 3, 4];
+        calculator.Add(summands); // capturedSummands captures copy of summands
+        summands.RemoveAt(0);
+        calculator.Add(summands); // capturedSummands captures copy of summands
+
+        // Assert
+        // both pass:
+        capturedSummands.Values[1].Should().BeEquivalentTo(2, 3, 4);
+        capturedSummands.Values[0].Should().BeEquivalentTo(1, 2, 3, 4);
+        //// --8<-- [end:CaptureCopiedMutatedList]
+    }
+
+    [Fact]
+    public void CaptureCopiedMutatedListToNewType()
+    {
+        //// --8<-- [start:CaptureCopiedMutatedListToNewType]
+        // Arrange
+        var capturedSummands = A.Captured<IEnumerable<int>, string>(
+            l => string.Join(" + ", l));

```

Comment:
This is the only part of the API that I'm not entirely happy with. Having to explicitly specify the type of the frozen argument is annoying. I think we could take advantage of type inference like this:

```csharp
        var capturedSummands = A.Captured<IEnumerable<int>>()
            .Freeze(l => string.Join(" + ", l));
```

`Freeze` (or whatever name we pick) would create a new instance of `Captured<TArgument, TCapture>`, based on the first `Captured<T>` (which would just be discarded).

What do you think?

Reply:
I sympathize with the annoying bonus typeparam. I'm not sold on the `.Freeze`, though. 
Partly because I think it's a _little_ less discoverable (I figure the second constructor will show up in intellisense more strongly, but maybe people will notice the `.Freeze` as well). But I also worry a little about a type that has `_`/`Ignored`/`That` **and** `Values`/`GetLastValue` hanging off it. We're mixing responsibilities. Maybe I worry a little.
Lemme sleep on it.

Reply:
Yeah, I don't love having an extra member that could be used in call specifications... But we have `Values`/`GetLastValue` anyway, which have the same problem. Maybe an analyzer for this would be useful (unless we can detect that we're in a call specification, and throw an exception in that case, but I suspect it's not that simple)

Reply:
Trying out the scenarios in the IDE.

Your `FrozenBy` scenario:

https://github.com/FakeItEasy/FakeItEasy/assets/3275797/077b68e6-a4bb-4056-918a-ca375950c946

The current overloads scenario:

https://github.com/FakeItEasy/FakeItEasy/assets/3275797/de82a532-377b-48f5-bed6-db12536cca30

Still to me the earlier IntelliSense feedback feels better, but I'll roll it around a little more.

Reply:
Somehow I missed your reply here.
To be honest, I'm not convinced. In my experience Intellisense isn't very helpful when picking between generic methods with different arity. Or maybe it's just in Rider.
Even if it's slightly worse in terms of discoverability, I'd pick the comfort of not having to specify the type explicitly every time.

Reply:
Also, I'd argue `.FrozenBy` is no less discoverable than, say, `.Returns` or `.MustHaveHappened`...

Reply:
>  I'd pick the comfort of not having to specify the type explicitly every time.

Every time someone captures. And wants to provide a "freezer" that changes the type of the captured argument. (and probably the new type is something simple like `string`)

> Also, I'd argue `.FrozenBy` is no less discoverable than, say, `.Returns` or `.MustHaveHappened`...

And finding those is enough of a problem that we added Diagnostic <html>
FakeItEasy0001: Unused call specification. 😄

But, I do like that `.FrozenBy` is a named operation rather than a random `Func` hanging out in the constructor. I'll whip a commit. Complain soon (or later) if you don't like `FrozenBy`.

Reply:
> Complain soon (or later) if you don't like `FrozenBy`.

I don't mind `FrozenBy`